### DISCUSSION TITLE
Zepp OS 3.0 Preview Documentation Updated with Heavyweight Features for File Downloading, File Transferring, and Image Conversion

### DISCUSSION THREAD

**Discussion by MyoungXUE** [zepp-health/zeppos-forum#217] [Announcements]
The three features mentioned in the title were very much in demand in the community, and in Zepp OS 3.0 we've implemented these features.

- Download File

The ability to download files from the web to the Side Service.

- File Transfer

Bi-directional file transfer between "Side Service" and "Device App".

- Image Conversion

Convert images to formats supported by the "Device App" and display them using the `IMG` UI widget.

**Comment by melianmiko** [comment]
That's very cool, but.... Only 3.0? 

File Download & Image Convertion are features that will be added to Side Service, and looks like they don't require modifying decice firmware. So, why this two features won't be available for older smart watches?

File transfer we could implement by themselves, with writing own MedsageBuilder with binary transfer support.

**Reply by MyoungXUE** [reply]
You're absolutely right.

We integrated the file transfer module into the framework in Zepp OS 3.0, eliminating the need to rely on externally introduced files, which is a difference from the current implementation and further simplifies the cost of use for developers. That's why we're claiming that this full set of features will be available in 3.0.

For developers capable of implementing file transfer protocols, it is possible to implement this set of features on 2.x systems.

**Reply by melianmiko** [reply]
Yes, of course it can be implemented. I already have a prototype with that features, but it uses my server for image processing, due to limitations of Side-Service Fetch API.

Due to current implementation of Fetch API in Side-Service, which didn't provide a way to get binary file contents, it works only with translating file content to base64, which is ugly and very slow...

So would be great if file download & image convert API will be available for older devices too. 
It will highly reduce network traffic usage (PNG images size was 2-4x smaller than ZeppOS-ready TGA), reduce my server CPU usage and of couse increase processing speed (due to no need to convert from/to base64).

**Reply by MyoungXUE** [reply]
This set of features is planned as new to Zepp OS 3.0 and has not been tested on Zepp OS 2.x devices, so there is no official guarantee of its usability on older devices.

Theoretically, there is no restriction on the framework version of Side Service for the time being, and the upgraded features may be available for Mini Programs with API_LEVEL 3.0 or below.

**Reply by melianmiko** [reply]
> there is no restriction on the framework version

That's all what I need, thank you)
